Crash on Huawei Android 7 / photoShoot

```
Heute, 20:37 in der App-Version 1030104
Huawei HUAWEI MediaPad T3 10 (HWAGS-Q), Android 7.0
Bericht 1 von 1
java.lang.RuntimeException: 
  at android.app.ActivityThread.performLaunchActivity (ActivityThread.java:2804)
  at android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:2875)
  at android.app.ActivityThread.-wrap12 (ActivityThread.java)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1578)
  at android.os.Handler.dispatchMessage (Handler.java:105)
  at android.os.Looper.loop (Looper.java:156)
  at android.app.ActivityThread.main (ActivityThread.java:6623)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:942)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:832)
Caused by: java.lang.IllegalArgumentException: 
  at androidx.core.content.FileProvider$SimplePathStrategy.getUriForFile (FileProvider.java:739)
  at androidx.core.content.FileProvider.getUriForFile (FileProvider.java:418)
  at com.nutomic.syncthingandroid.activities.PhotoShootActivity.openCameraIntent (PhotoShootActivity.java:167)
  at com.nutomic.syncthingandroid.activities.PhotoShootActivity.onCreate (PhotoShootActivity.java:77)
  at android.app.Activity.performCreate (Activity.java:6910)
  at android.app.Instrumentation.callActivityOnCreate (Instrumentation.java:1123)
  at android.app.ActivityThread.performLaunchActivity (ActivityThread.java:2757)
```

--- a/app/src/main/java/com/nutomic/syncthingandroid/activities/PhotoShootActivity.java
+++ b/app/src/main/java/com/nutomic/syncthingandroid/activities/PhotoShootActivity.java
@@ -158,6 +158,10 @@ public class PhotoShootActivity extends AppCompatActivity {
         File photoFile = null;
         try {
             photoFile = createImageFile();
+            if (photoFile == null) {
+                Log.e(TAG, "openCameraIntent: photoFile == null");
+                return;
+            }
         } catch (IOException ex) {
             Log.e(TAG, "Error occurred while creating the temp image file");
             return;
@@ -183,7 +187,11 @@ public class PhotoShootActivity extends AppCompatActivity {
                           Locale.getDefault()).format(new Date());
         String imageFileName = "IMG_" + timeStamp + "_";
         File storageDir =
-                    getExternalFilesDir(Environment.DIRECTORY_PICTURES);
+                    this.getExternalFilesDir(this, Environment.DIRECTORY_PICTURES);
+        if (storageDir == null) {
+            Log.e(TAG, "createImageFile: storageDir == null");
+            return null;
+        }
         File image = File.createTempFile(
                         imageFileName,  /* prefix */
                         ".jpg",         /* suffix */
@@ -275,5 +283,17 @@ public class PhotoShootActivity extends AppCompatActivity {
     private void updateButtons() {
         mBtnGrantCameraPerm.setVisibility(haveCameraPermission() ? View.GONE : View.VISIBLE);
         mBtnGrantStoragePerm.setVisibility(haveStoragePermission() ? View.GONE : View.VISIBLE);
+    }
+
+    public static File getExternalFilesDir(Context context, String type) {
+    	// There is a bug on Huawei devices running Android 7, which returns the wrong external path.
+        // See https://github.com/Catfriend1/syncthing-android/issues/541
+    	// ... and: https://stackoverflow.com/questions/39895579/fileprovider-error-onhuawei-devices
+        File[] externalFilesDirs = ContextCompat.getExternalFilesDirs(context, type);
+        if (externalFilesDirs.length > 0) {
+            return externalFilesDirs[0];
+        } else {
+            return null;
+        }
     }
 }

